presto-spi/src/main/java/io/prestosql/spi/block/MapHashTables.java

```diff
@@ -15,23 +15,128 @@
+        }
+        return hashTables;
+    }
+
+    /**
+     * Returns the raw hash tables, if they have been built.  The raw hash tables must not be modified.
+     */
+    Optional<int[]> tryGet()
+    {
+        return Optional.ofNullable(hashTables);
+    }
+
+    synchronized void growHashTables(int newSize)
     {
+        int[] hashTables = this.hashTables;
+        if (hashTables == null) {
+            throw new IllegalStateException("hashTables not set");
+        }
+        if (newSize < hashTables.length) {
+            throw new IllegalArgumentException("hashTables size does not match expectedEntryCount");
+        }
+        int[] newRawHashTables = Arrays.copyOf(hashTables, newSize);
+        Arrays.fill(newRawHashTables, hashTables.length, newSize, -1);
+        this.hashTables = newRawHashTables;
+    }
+
+    synchronized void buildAllHashTables(Block rawKeyBlock, int[] offsets, @Nullable boolean[] mapIsNull)
+    {
+        if (hashTables != null) {
+            return;
+        }
+
+        int[] hashTables = new int[rawKeyBlock.getPositionCount() * HASH_MULTIPLIER];
+        Arrays.fill(hashTables, -1);
+
+        int hashTableCount = offsets.length - 1;
+        for (int i = 0; i < hashTableCount; i++) {
+            int keyOffset = offsets[i];
+            int keyCount = offsets[i + 1] - keyOffset;
+            if (keyCount < 0) {
+                throw new IllegalArgumentException(format("Offset is not monotonically ascending. offsets[%s]=%s, offsets[%s]=%s", i, offsets[i], i + 1, offsets[i + 1]));
+            }
+            if (mapIsNull != null && mapIsNull[i] && keyCount != 0) {
+                throw new IllegalArgumentException("A null map must have zero entries");
+            }
+            buildHashTableInternal(rawKeyBlock, keyOffset, keyCount, hashTables);
+        }
+        this.hashTables = hashTables;
+    }
+
+    synchronized void buildHashTable(Block keyBlock, int keyOffset, int keyCount)
+    {
+        int[] hashTables = this.hashTables;
+        if (hashTables == null) {
+            throw new IllegalStateException("hashTables not set");
+        }
+
+        buildHashTableInternal(keyBlock, keyOffset, keyCount, hashTables);
+        this.hashTables = hashTables;
+    }
+
+    private void buildHashTableInternal(Block keyBlock, int keyOffset, int keyCount, int[] hashTables)
+    {
+        int hashTableOffset = keyOffset * HASH_MULTIPLIER;
+        int hashTableSize = keyCount * HASH_MULTIPLIER;
         for (int i = 0; i < keyCount; i++) {
-            int hash = getHashPosition(mapType, keyBlock, keyOffset + i, hashTableSize);
+            int hash = getHashPosition(keyBlock, keyOffset + i, hashTableSize);
             while (true) {
-                if (outputHashTable[hashTableOffset + hash] == -1) {
-                    outputHashTable[hashTableOffset + hash] = i;
+                if (hashTables[hashTableOffset + hash] == -1) {
```

Comment:
Can this be read by some other thread *while* it's being built?

Reply:
No.  In the case of a normal Map, the table does not exist until after this method finishes.  In the case of a block builder map, the map is built as each entry is added, and the entry does not exist until the entry is closed.  Also block builders are not thread safe.